### ISSUE TITLE
System.Management.Automation.Signature type definition error

### ISSUE THREAD

**Issue by PatrickOnGit** [PKISolutions/PSPKI#83]

**Comment by PatrickOnGit** [comment]
The type definition of System.Management.Automation.Signature in PSPKI.Types.ps1xml contains two incorrect lines which prevents returning SigningTime:
Line 90
 `$_.Path,` must be replaced with `$this.path`,
Line 111
`$CounterSignerInfos` must be commented out our removed

**Comment by Crypt32** [comment]
I checked these fixes, however it still doesn't work. When reading timestamping signature, `$SignedCms.SignerInfos[0].CounterSignerInfos` returns null. I verified with ASN.1 editor and timestamp exists in signed attrbiutes. For some reasons `SignedCms` class doesn't recognize it.

**Comment by Crypt32** [comment]
I found timestamp value in unauthenticated attributes which contains embedded PKCS#7 structure. I need to investigate when it was changed and adjust the code.

**Comment by PatrickOnGit** [comment]
Hi Vadims

Sorry for not responding to your first response.

The document https://docs.microsoft.com/en-us/windows/win32/seccrypto/time-stamping-authenticode-signatures still mentions the following:

"An example of an authenticated attribute is the signing time (OID 1.2.840.113549.9.5) because it is part of what the time stamp service signs."

I guess the ASN structure itself does not change after it is signed. So the structure would depend on the signer, not the one validating the time stamp. Am I correct?

If authenticated as well as unauthenticated attributes are valid you store timestamp, wouldn't it be most easy to check both locations, assuming only one contains data or authenticated is preferred over unauthenticated.

Btw. The function dreams only to work with signed PowerShell. It seems TS on exe and alike can not be made visible with this code.

Best regards
Patrick

________________________________
Von: Vadims Podans <notifications@github.com>
Gesendet: Sonntag, 11. August 2019 20:25
An: PKISolutions/PSPKI
Cc: PatrickOnGit; Author
Betreff: Re: [PKISolutions/PSPKI] System.Management.Automation.Signature type definition error (#83)

I found timestamp value in unauthenticated attributes which contains embedded PKCS#7 structure. I need to investigate when it was changed and adjust the code.

**Comment by Crypt32** [comment]
In short, there are two counter signing (timestamping) formats. One uses RFC3161 format and another uses PKCS#9 format. I'm working on it, to implement both formats.

PKCS#9 seems to be a simple unauthenticated attribute, while RFC3161 is a whole PKCS#7 message embedded as unauthenticated attribute of signature's PKCS#7 message.
Here is how it looks like RFC3161 timestamping:
![sig-rfc3161](https://user-images.githubusercontent.com/6384119/62884553-872d6800-bd3f-11e9-8c52-9caefef636a6.PNG)

and PKCS#9:
![sig-pkcs9](https://user-images.githubusercontent.com/6384119/62884487-59e0ba00-bd3f-11e9-91ed-ccda7f9e1991.PNG)

**Comment by PatrickOnGit** [comment]
Good morning Vadims

Thank you very much for the update.
Would it be possible to implement this without re-reading the files? Get-AuthenticodeSignature already contains the signer certificate. Though I'm not sure if this contains the timestamp. 
And, it would be nice if it would work with all authenticode signed files (PoSh, exe, msi, etc.). 

Best regards
Patrick
________________________________
Von: Vadims Podans <notifications@github.com>
Gesendet: Montag, 12. August 2019 19:28
An: PKISolutions/PSPKI
Cc: PatrickOnGit; Author
Betreff: Re: [PKISolutions/PSPKI] System.Management.Automation.Signature type definition error (#83)

In short, there are two counter signing (timestamping) formats. One uses RFC3161 format and another uses PKCS#9 format. I'm working on it, to implement both formats.

PKCS#9 seems to be a simple unauthenticated attribute, while RFC3161 is a whole PKCS#7 message embedded as unauthenticated attribute of signature's PKCS#7 message.
Here is how it looks like RFC3161 timestamping:

and PKCS#9:

**Comment by Crypt32** [comment]
> Would it be possible to implement this without re-reading the files?

no. Get-AuthenticodeSignature doesn't give me access to signature.

> Though I'm not sure if this contains the timestamp

of course, no. Signing certificate itself gives me zero value. I need to extract the whole signature in order to find timestamp info.

> And, it would be nice if it would work with all authenticode signed files (PoSh, exe, msi, etc.).

Signature format doesn't depend on file extension. The only difference is how files are timestamped and it depends on tools used to sign the file.

**Comment by Crypt32** [comment]
fixed. This fix includes related bug in custom PKCS SignerInfo decoder: https://github.com/PKISolutions/pkix.net/commit/dcae01b647b0249ff2536d275bf813b3c3ff1c53

I've tested with several different files, seems to be working. If you wish, I can compile you draft release for testing on your side.

**Comment by PatrickOnGit** [comment]
Hello Vadims

That was quick!
I would be happy to test.

Best regards
Patrick


________________________________
Von: Vadims Podans <notifications@github.com>
Gesendet: Dienstag, 13. August 2019 09:50
An: PKISolutions/PSPKI
Cc: PatrickOnGit; Author
Betreff: Re: [PKISolutions/PSPKI] System.Management.Automation.Signature type definition error (#83)

fixed. This fix includes related bug in custom PKCS SignerInfo decoder: PKISolutions/pkix.net@dcae01b

I've tested with several different files, seems to be working. If you wish, I can compile you draft release for testing on your side.

**Comment by Crypt32** [comment]
Here it is. Please, let me know if it works or doesn't work for you.
[PSPKI.zip](https://github.com/PKISolutions/PSPKI/files/3498823/PSPKI.zip)

**Comment by PatrickOnGit** [comment]
Hello Vadims

 

Thank you for the updated module.

Unfortunately it doesn’t seem to work.

I made sure that I opened a new session without any trace of PSPKI, imported the new version of the Module.

The type is loaded 

Get-TypeData  System.Management.Automation.Signature | Select -Expand Members

 

Key         Value

---         -----

SigningTime System.Management.Automation.Runspaces.ScriptPropertyData

 

But I do not get the SigningTime  back:

PS C:\Users\Patrick > Get-AuthenticodeSignature 'C:\Program Files\WindowsPowerShell\Modules\PSPKI\3.4.1.0\PSPKI.psd1' | Select *

 

SigningTime            :

SignerCertificate      : [Subject]

                           CN=Sysadmins LV IK, O=Sysadmins LV IK, L=Riga, C=LV

…

TimeStamperCertificate : [Subject]

                           CN=DigiCert Timestamp Responder, O=DigiCert, C=US

 

Status                 : Valid

StatusMessage          : Signature verified.

Path                   : C:\Program Files\WindowsPowerShell\Modules\PSPKI\3.4.1.0\PSPKI.psd1

SignatureType          : Authenticode

IsOSBinary             : False

 

When I run the code directly the object returned in line 129 does not contain a signed attribute as expected in 

 

PS C:\Users\Patrick > $Nested = New-Object SysadminsLV.PKI.Cryptography.Pkcs.PkcsSignerInfo (,$counter.Values[0].RawData)

PS C:\Users\Patrick > $Nested

 

 

Version                   : 1

Issuer                    : SysadminsLV.PKI.Cryptography.Pkcs.SubjectIdentifier2

Certificate               :

HashAlgorithm             : Signature Algorithm:

                                Algorithm ObjectId: sha1 (1.3.14.3.2.26)

                                Algorithm Parameters:

                            05 00

 

EncryptedHashAlgorithm    : Signature Algorithm:

                                Algorithm ObjectId: RSA (1.2.840.113549.1.1.1)

                                Algorithm Parameters:

                            05 00

 

EncryptedHash             : {88, 19, 84, 75...}

AuthenticatedAttributes   : {0, 0, 0}

UnauthenticatedAttributes : {}

 

Best regards

Patrick

**Comment by Crypt32** [comment]
My bad. I'm calling an incorrect property. In `PSPKI.Types.ps1xml` file on line 131 replace `SignedAttributes` with `AuthenticatedAttributes`.
https://github.com/PKISolutions/PSPKI/commit/7e955e0ad3b4a778cac7426f2eeb35bb6ab5a630

this should work

**Comment by PatrickOnGit** [comment]
Changing line 131 did the trick for PKCS#9.

 

But the RFC3161 part is not working as well.

 

 

Best regards

Patrick

**Comment by Crypt32** [comment]
Can you give me an example of non-working RFC3161?

**Comment by PatrickOnGit** [comment]
I changed the code slightly to get some output even if the script return preliminary and looped through system32 folder:

 

ls 'C:\Windows\System32\'  -File -ErrorAction SilentlyContinue | %{ try{ Get-AuthenticodeSignature -FilePath $_.Fullname -ErrorAction SilentlyContinue } catch {}} | Where Status -eq Valid | Select SigningTime, Path | Export-Csv -NoTypeInformation C:\Temp\SigningTimeTest.csv

 

Because I only return files with a valid signature, you should find examples of files not working. Here are some:

 

Found signature, assume PKCS9 and successfully return timestamp:

                C:\Windows\System32\nvdispco6438875.dll

 

Found signature, assume RFC3161 and successfully return timestamp:

C:\Windows\System32\vulkan-1.dll

 

Found signature, assume RFC3161 but does not return timestamp:     

C:\Windows\System32\aeinv.dll

 

CryptQueryObject Failed, assumes there is no signature at all though this is a signed file: 

C:\Windows\System32\AgentService.exe

 

Best regards

Patrick

**Comment by Crypt32** [comment]
Your test isn't much valid. Failed entries are failed for a reason. Most files in system32 folder use catalog signing. You can check it by looking at `SignatureType` property. My code is supposed to work with authenticode signature only. Catalog signing is not supported and it is not in my roadmap. Check it again against authenticode signed files.

**Comment by PatrickOnGit** [comment]
Hello Vadims

 

Thanks for your response and good to know. It isn’t obvious for me as user that the code only supports Authenticode and not Catalog files. But fair enough. 

Just saying a “Verbose” output or some kind of error message (I guess you don’t want to break anything just because of that) would help to make it more visible that only Authenticode signed code is in scope.

 

Ok, slight update:

ls 'C:\Windows\System32\'  -File -ErrorAction SilentlyContinue | %{ try{ Get-AuthenticodeSignature -FilePath $_.Fullname -ErrorAction SilentlyContinue } catch {}} | Where SignatureType -eq 'Authenticode' | Where Status -eq Valid | Select SigningTime, Path | Export-Csv -NoTypeInformation C:\Temp\SigningTimeTest.csv

 

Not Working RFC3161:

C:\Windows\System32\igfxCoIn_v4358.dll

Working RFC3161:

                C:\Windows\System32\tphkcoinst.dll

 

And I also scanned PowerShell modules in Program Files (See SignintTimeTest2). This has more examples of signatures assumed to be RFC3161 but not getting the SigningTime.

 

Another note, the type extension crashes the PoSh console when scanning the PowerShell Modules folder although I put a try catch in calling Get-Authenticate as well as putting a try catch around the scriptblock in ps1xml.

 

Best regards

Patrick

**Comment by Crypt32** [comment]
> It isn’t obvious for me as user that the code only supports Authenticode

it should be based on cmdlet name: Get-**Authenticode**Signature. This is what I'm focusing on. The fact that cmdlet supports something else is a different question.

> Another note, the type extension crashes the PoSh console when scanning the PowerShell Modules folder

this issue should be addressed to PS development team.

**Comment by PatrickOnGit** [comment]
Hello Vadims

... this issue should be addressed to PS development team.

The cmdlet Get-AuthenticodeSignature does NOT crash when pspki is NOT loaded. It only crashes when pspki is imported.
So I guess it is an issue of the type extension.

Best regards
Patrick

________________________________
Von: Vadims Podans <notifications@github.com>
Gesendet: Mittwoch, 14. August 2019 22:31
An: PKISolutions/PSPKI
Cc: PatrickOnGit; Author
Betreff: Re: [PKISolutions/PSPKI] System.Management.Automation.Signature type definition error (#83)

> It isn’t obvious for me as user that the code only supports Authenticode

it should be based on cmdlet name: Get-AuthenticodeSignature. This is what I'm focusing on. The fact that cmdlet supports something else is a different question.
>
> Another note, the type extension crashes the PoSh console when scanning the PowerShell Modules folder

this issue should be addressed to PS development team.

**Comment by Crypt32** [comment]
> The cmdlet Get-AuthenticodeSignature does NOT crash when pspki is NOT loaded.

I understand. In fact, the call to unmanaged function (CryptQueryObject) crashes the console in certain cases. What cases? Have no idea, because in most cases, the code works normally, so I believe the call is correct and the problem is somewhere either in `CryptQueryObject` function or in PS itself.

PowerShell hides all managed exceptions raised in ps1xml file and silently continue. However, PowerShell still has issues when calling unmanaged functions and PowerShell don't handle them. Often it is because of memory access violation or memory corruption. That's from my personal experience with Win32 in PS.

**Comment by PatrickOnGit** [comment]
Hello Vadims

 

Thank you for your feedback.

The original issue, that not all Authenticode signed signature timestamps can be read, here is another example, see attached files.

Use password as password.

 

This one works:

Get-AuthenticodeSignature C:\Windows\System32\tphkcoinst.dll | ft SigningTime, Status, StatusMessage, Path, SignatureType, IsOSBinary

 

SigningTime                   Status StatusMessage       Path                               SignatureType IsOSBinary

-----------                   ------ -------------       ----                               ------------- ----------

RFC3161 - 10/16/2017 07:40:06  Valid Signature verified. C:\Windows\System32\tphkcoinst.dll  Authenticode      False

 

This one doesn’t

Get-AuthenticodeSignature C:\Windows\System32\igfxCoIn_v4358.dll | ft SigningTime, Status, StatusMessage, Path, SignatureType, IsOSBinary

 

SigningTime Status StatusMessage       Path                                   SignatureType IsOSBinary

----------- ------ -------------       ----                                   ------------- ----------

RFC3161 -    Valid Signature verified. C:\Windows\System32\igfxCoIn_v4358.dll  Authenticode      False

 

Both files have two signatures, each timestamped (using explorer to view).

 

Thank you for your support

Best regards

Patrick

**Comment by PatrickOnGit** [comment]
Hello Vadims

have you been able to test with the two files I sent? Did you fix the problem?

best regards
Patrick

**Comment by Crypt32** [comment]
Partially, it is fixed. However, I haven't had a chance to look at last items. I think, original issue is fixed and thread went into a bit another issue. So let's start this in separate thread.

**Comment by PatrickOnGit** [comment]
Hello Vadims

ok, thank you

**Comment by Crypt32** [comment]
Just copy/paste items that still don't work and specify if failing items are authenticode-signed or use other signing schemes.

**Comment by PatrickOnGit** [comment]
Here are the example files. Both use authenticode


[No timeStamp returned - VMware.DeployAutomation - password.zip](https://github.com/PKISolutions/PSPKI/files/3544399/No.timeStamp.returned.-.VMware.DeployAutomation.-.password.zip)
[No timeStamp returned - Plaster - password.zip](https://github.com/PKISolutions/PSPKI/files/3544401/No.timeStamp.returned.-.Plaster.-.password.zip)

**Comment by Crypt32** [comment]
Please, move this to new issue thread.
